airflow/api/common/experimental/delete_dag.py

```diff
@@ -45,6 +49,11 @@ def delete_dag(dag_id: str, keep_records_in_log: bool = True, session=None) -> i
         raise DagFileExists("Dag id {} is still in DagBag. "
                             "Remove the DAG file first: {}".format(dag_id, dag.fileloc))
 
+    # Scheduler removes DAGs without files from serialized_dag table every dag_dir_list_interval.
+    # There may be a lag, so explicitly removes serialized DAG here.
+    if DAGCACHED_ENABLED and SerializedDagModel.has_dag(dag_id):
+        SerializedDagModel.remove_dag(dag_id)
```

Comment:
```suggestion
        SerializedDagModel.remove_dag(dag_id, session=session)
```

May need to extend that method to take a session argument, but this way it all happens inside the same DB txn.

Reply:
Updated in https://github.com/apache/airflow/pull/5743/commits/b814f8dfd9448ee3ceef2722c7f0291d8a680700